sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala

```diff
@@ -202,3 +203,14 @@ object Unions {
     }
   }
 }
+
+/**
+ * Extractor for retrieving Int value.
+ */
+object IntegerIndex {
+  def unapply(a: Any): Option[Int] = a match {
+    case Literal(a: Int, IntegerType) => Some(a)
+    case UnaryMinus(IntegerLiteral(v)) => Some(-v)
```

Comment:
This line is used for catching the illegal case:
```scala
      sql("SELECT * FROM testData2 ORDER BY -1 DESC, b ASC").collect()
```

I plan to keep it untouched in the PR. Thanks!